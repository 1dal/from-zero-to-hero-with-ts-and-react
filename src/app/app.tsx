import React, { Component } from 'react'

export type AppProps = {
  hide?: boolean
}
export type AppState = typeof initialState
const initialState = {
  on: false,
  input: '',
  mainColor: 'blue' as 'blue' | 'red',
  lifeCycle: ''
}
export class App extends Component<AppProps, AppState> {
  state = initialState

  private readonly toggleOnState = () =>
    this.setState((state) => ({ on: !state.on }))

  handleStrings(value: string) {
    return value === 'Hello World' ? true : false
  }
  componentWillReceiveProps() {
    this.setState({ lifeCycle: 'componentWillReceiveProps' })
  }
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Welcome to React</h1>
          <h3 className={this.state.mainColor}>Everyone is Welcome!</h3>
        </header>
        <p className="App-intro">Hello</p>
        <p className="button-state">{this.state.on ? 'Yes!' : 'No!'}</p>
        <button onClick={this.toggleOnState}>Click</button>
        <h2>{this.state.input}</h2>
        <input
          onChange={(e) => this.setState({ input: e.currentTarget.value })}
          type="text"
        />
        <p className="lifeCycle">{this.state.lifeCycle}</p>
      </div>
    )
  }

  componentDidMount() {
    // console.log('CDM!')
    this.setState({ lifeCycle: 'componentDidMount' })
  }
}

export type LinkProps = {
  address: string
  hide?: boolean
}
export class Link extends Component<LinkProps> {
  render() {
    return this.props.hide ? null : <a href={this.props.address}>Click</a>
  }
}
