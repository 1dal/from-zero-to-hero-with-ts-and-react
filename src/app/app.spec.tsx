import React, { AnchorHTMLAttributes } from 'react'
import { configure, shallow, ShallowWrapper, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { render, unmountComponentAtNode } from 'react-dom'
import toJson from 'enzyme-to-json'

import { App, Link, AppProps, AppState } from './app'

configure({ adapter: new Adapter() })

type LinkProps = import('./app').LinkProps

describe(`shallow rendering`, () => {
  describe(`<${App.name}/>`, () => {
    it(`should render App`, () => {
      const wrapper = shallow(<App />, { disableLifecycleMethods: true })
      // expect(wrapper.contains('hello')).toBeTruthy()
      // console.log(wrapper.debug())
    })

    it(`should contain 1 p element`, () => {
      const wrapper = shallow(<App />)
      expect(wrapper.find({ className: 'App-intro' })).toHaveLength(1)
    })
    it(`should match the snapshot`, () => {
      const tree = shallow(<App />)
      expect(toJson(tree)).toMatchSnapshot()
    })

    // it(`renders without crashing`, () => {
    //   const root = document.createElement('div')
    //   render(<App />, root)
    //   unmountComponentAtNode(root)
    // })

    it(`should changes p text on button click`, () => {
      const wrapper = shallow(<App />)
      const button = wrapper.find('button')

      expect(wrapper.find('.button-state').text()).toBe('No!')

      button.simulate('click')

      expect(wrapper.find('.button-state').text()).toBe('Yes!')
    })

    it(`should change title text on input change`, () => {
      const wrapper = shallow(<App />)
      const input = wrapper.find('input')

      expect(wrapper.find('h2').text()).toBe('')

      input.simulate('change', { currentTarget: { value: 'Tyler' } })

      expect(wrapper.find('h2').text()).toBe('Tyler')
    })

    it(`should update className with new State`, () => {
      const wrapper = shallow<{}, import('./app').AppState>(<App />)

      expect(wrapper.find('.blue').length).toBe(1)
      expect(wrapper.find('.red').length).toBe(0)

      wrapper.setState({ mainColor: 'red' })

      expect(wrapper.find('.blue').length).toBe(0)
      expect(wrapper.find('.red').length).toBe(1)
    })

    it(`should call componentDidMount`, () => {
      jest.spyOn(App.prototype, 'componentDidMount')
      const wrapper = shallow(<App />)

      expect(App.prototype.componentDidMount).toHaveBeenCalled()
      expect(wrapper.find('.lifeCycle').text()).toBe('componentDidMount')
    })

    it(`should call componentWillReceiveProps via setProps`, () => {
      const componentWillReceivePropsSpy = jest.spyOn(
        App.prototype,
        'componentWillReceiveProps'
      )
      const wrapper = shallow<AppProps, AppState>(<App />)
      wrapper.setProps({ hide: false })

      expect(componentWillReceivePropsSpy.mock.calls.length).toBe(1)
      expect(wrapper.find('.lifeCycle').text()).toBe(
        'componentWillReceiveProps'
      )
    })

    it(`should return correctly when handleStrings function is called`, () => {
      const wrapper = shallow<AppProps, AppState>(<App />)
      const instance: App = wrapper.instance() as App
      const trueReturn = instance.handleStrings('Hello World')
      const falseReturn = (wrapper.instance() as App).handleStrings('')

      expect(trueReturn).toBe(true)
      expect(falseReturn).toBe(false)
    })
  })

  describe(`<Link/>`, () => {
    it(`should link`, () => {
      const el = <Link address="www.google.com" />
      const wrapper = shallow<LinkProps>(<Link address="www.google.com" />)

      expect(wrapper.instance().props.address).toBe('www.google.com')
    })
    it(`should render href correctly on a tag`, () => {
      const wrapper = shallow<LinkProps & AnchorHTMLAttributes<{}>>(
        <Link address="www.google.com" />
      )

      expect(wrapper.props().href).toBe('www.google.com')
    })

    it(`should return null with true hide prop`, () => {
      const wrapper = shallow<LinkProps>(<Link address="''" hide={false} />)
      expect(wrapper.find('a')).toHaveLength(1)

      wrapper.setProps({ hide: true })
      expect(wrapper.get(0)).toBeNull()
    })
  })
})

describe(`mount rendering`, () => {
  it(`should contain correct text in h1`, () => {
    const wrapper = mount(<App />)
    expect(wrapper.find({ className: 'App-intro' })).toHaveLength(1)
  })

  it(`should match the snapshot`, () => {
    const tree = mount(<App />)
    expect(toJson(tree)).toMatchSnapshot()
    // tree.unmount()
  })
})
