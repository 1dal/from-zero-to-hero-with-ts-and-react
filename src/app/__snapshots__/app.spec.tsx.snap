// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount rendering should match the snapshot 1`] = `
<App>
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <h1
        className="App-title"
      >
        Welcome to React
      </h1>
      <h3
        className="blue"
      >
        Everyone is Welcome!
      </h3>
    </header>
    <p
      className="App-intro"
    >
      Hello
    </p>
    <p
      className="button-state"
    >
      No!
    </p>
    <button
      onClick={[Function]}
    >
      Click
    </button>
    <h2 />
    <input
      onChange={[Function]}
      type="text"
    />
    <p
      className="lifeCycle"
    >
      componentDidMount
    </p>
  </div>
</App>
`;

exports[`shallow rendering <App/> should match the snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
    <h3
      className="blue"
    >
      Everyone is Welcome!
    </h3>
  </header>
  <p
    className="App-intro"
  >
    Hello
  </p>
  <p
    className="button-state"
  >
    No!
  </p>
  <button
    onClick={[Function]}
  >
    Click
  </button>
  <h2 />
  <input
    onChange={[Function]}
    type="text"
  />
  <p
    className="lifeCycle"
  >
    componentDidMount
  </p>
</div>
`;
